packages/apputils/src/vdom.ts

```diff
@@ -127,6 +160,112 @@ export class ReactElementWidget extends VDomRenderer<any> {
   private _es: Array<React.ReactElement<any>> | React.ReactElement<any> | null;
 }
 
+/**
+ * Props for the UseSignal component
+ */
+export interface IUseSignalProps<SENDER, ARGS> {
+  /**
+   * Phosphor signal to connect to.
+   */
+
+  signal: ISignal<SENDER, ARGS>;
+  /**
+   * Initial value to use for the sender, used before the signal emits a value.
+   * If not provided, initial sender will be undefined
+   */
+  sender?: SENDER;
+  /**
+   * Initial value to use for the args, used before the signal emits a value.
+   * If not provided, initial args will be undefined.
+   */
+  args?: ARGS;
+  /**
+   * Function mapping the last signal value or inital values to an element to render.
+   */
+
+  children: (sender?: SENDER, args?: ARGS) => JSX.Element;
+  /**
+   * Given the last signal value, should return whether to update the state or not.
+   *
+   * The default unconditionally returns `true`, so you only have to override if you want
+   * to skip some updates.
+   */
+
+  shouldUpdate?: (sender: SENDER, args: ARGS) => boolean;
+}
+
+/**
+ * State for the UseSignal component
+ */
+export interface IUseSignalState<SENDER, ARGS> {
+  value: [SENDER, ARGS];
+}
+
+/**
+ * UseSignal provides a way to hook up a Phosphor signal to a React element,
+ * so that the element is re-rendered every time the signal fires.
+ *
+ * It is implemented through the "render props" technique, using the `children`
+ * prop as a function to render, so that it can be used either as a prop or as a child
+ * of this element
+ * https://reactjs.org/docs/render-props.html
+ *
+ *
+ * Example as child:
+ *
+ * ```
+ * function LiveButton(isActiveSignal: ISignal<any, boolean>) {
+ *  return (
+ *    <UseSignal signal={isActiveSignal} initialArgs={True}>
+ *     {(_, isActive) => <Button isActive={isActive}>}
+ *    </UseSignal>
+ *  )
+ * }
+ * ```
+ *
+ * Example as prop:
+ *
+ * ```
+ * function LiveButton(isActiveSignal: ISignal<any, boolean>) {
+ *  return (
+ *    <UseSignal
+ *      signal={isActiveSignal}
+ *      initialArgs={True}
+ *      children={(_, isActive) => <Button isActive={isActive}>}
+ *    />
+ *  )
+ * }
+ */
+export class UseSignal<SENDER, ARGS> extends React.Component<
```

Comment:
I like this pattern 👍👍

Should we consider allowing for multiple signals? My first through:

```jsx
<UseSignal signals=[{signal: props.session.kernelChanged, sender: props.session}, {signal: props.commands.commandChanged, sender: props.commands}]>
    ([session, commands]) => {
        // do something with session or commands
    }
</UseSignal>
````

Nevermind, this will callback when either type of signal... The React hook interface will serve the same purpose and clean it up even more 👍

Reply:
So you think we should or should not extend this for multiple signals?

I would opt for the more verbose and explicit form of nesting them currently:

```jsx
<UseSignal signal={props.session.kernelChanged} sender={props.session}>
  {(session, kernel) => (
    <UseSignal signal={props.commands.commandChanged} sender={props.commands}>
      {(commandRegistry, commandsChanged) => {
        // do something with kernel and commands
      }}
    </UseSignal>
  )}
</UseSignal>
```

Reply:
Agreed. Accepting an array of signals doesn't really work as it would callback whenever any signal is received (which is ok but I think your example is better).